odoo/odoo
# [IMP] mail: add mock server for common rpc

Improve mock server:

- add support for mocked `fetch`
- add support for `active_test`
- add support for x2m `in` in domains
- add support for default values computed from a function
- implement a more natural "next id" compute
- allow initial data without ids
- ensure write and x2m commands integrity
- improve bad data/bad commands error messages
- always warn for failing RPC, not only in debug mode
- fix all existing tests that had inconsistency data

Other changes done in mail (or dependents) that are not just related to tests:

- remove `direct_partner` from formatter result
  ->`correspondent` can be computed from other keys, especially `members`

- fix `livechat_visitor` convertData
  -> only process if there is value

- add `current_partner` and `current_user_id` as `init_messaging` result
  -> easier to mock than session

- remove usage of `need_moderation`
  -> that was just a search indirection to `moderation_status`

- adapt `partner_id` -> `res_partner_id` key in `_notification_format`
  -> to be consistent with field name

- add name in result of `mail_partner_format`
  -> sometimes display_name is not the same

- remove usage of `is_moderator`
  -> that was just an indirection to `moderation_channel_ids`

Enterprise counterpart: odoo/enterprise#11523
task-2287171


various notes:
- mockrpc -> mock server (except asserts)
- no initMessaging
- this.data instead of this.env.models create when possible (done some, out of scope to do all of them)
- records push and not override
- mock models override for all modules (not fully done, outside of mail / out of scope)
- mock web content/image route (was already the case)
- comment reason of each data (remove useless/default data) (checked and done a decent part, too long for the rest)
- only use ids above 10 in data (done when diff on line, too much work to check all of them)
- mentions improve tests to check main and extra suggestions (out of scope, but necessary to ensure mock is correct)
- data definition should use data and not env (current partnet etc when doing notif manually for example) (the problem is doing notif manually and not with the mock server, out of scope)
- mock server for channel message not returning message / not setting channel_ids in message if pending moderation (I don't remember what I meant, not important for existing tests)
- session should be set from data current partner etc (out of scope)
- review all mocks naming & docstring (if from models: _MockModelMethod)
- mark all require focus tests
- replace message_needaction_counter by actual messages/notifs
- message_unread_counter by actual messages/notifs (not done but I don't want to delay the PR further)
- always use _mockCreate / _mockWrite
- don't update data in place with formatters...
- try to remove bus trigger in test, replace by actual actions/rpc leading to result to ensure data integrity (out of scope)
